### PR TITLE
refactor(routing): update top level and some nested routes top React Router v6

### PR DESCRIPTION
**Pull request by huwshimi**
## Done

- Update the top level routes and the minimum number of child routes to work with React Router v6.

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Click through the top level and child routes for each section of the UI.
- Pay particular attention to the KVM routes as they've been modified the most.

## Fixes

Fixes: canonical-web-and-design/app-tribe#1045.

### CODE REVIEW

**Comment by huwshimi** [comment]
Apologies for the size, this was the smallest change I could make to start moving the routes over to RRv6.

File: src/app/Routes.tsx [hunk 1]
```diff
@@ -29,177 +31,215 @@ import Preferences from "app/preferences/views/Preferences";
...
     <Route
-      path={introURLs.index}
-      render={() => (
+      element={
         <ErrorBoundary>
           <Intro />
         </ErrorBoundary>
-      )}
+      }
+      path={`${introURLs.index}/*`}
```
Comment (huwshimi) [inline src/app/Routes.tsx:55]: Each top level route now has `/*` to capture its descendants.

File: src/app/Routes.tsx [hunk 2]
```diff
@@ -29,177 +31,215 @@ import Preferences from "app/preferences/views/Preferences";
...
+        <ErrorBoundary>
+          <SubnetsList />
+        </ErrorBoundary>
+      }
+      path={`${subnetsURLs.index}/*`}
+    />
+    <Route
+      element={
+        <ErrorBoundary>
+          <FabricDetails />
```
Comment (huwshimi) [inline src/app/Routes.tsx:188]: These components use to be in `Subnets.tsx` but now need to be in this file as they each have their own top level route e.g. /fabric/

File: src/app/base/components/node/NodeTestDetails/NodeTestDetails.test.tsx [hunk 3]
```diff
@@ -47,10 +47,12 @@ describe("NodeTestDetails", () => {
       <Provider store={store}>
         <MemoryRouter initialEntries={["/machine/abc123/testing/1/details"]}>
           <CompatRouter>
-            <Route
-              path="/machine/:id/testing/:scriptResultId/details"
-              render={() => <NodeTestDetails getReturnPath={getReturnPath} />}
-            />
+            <Routes>
```
Comment (huwshimi) [inline src/app/base/components/node/NodeTestDetails/NodeTestDetails.test.tsx:50]: A lot of components had to be updated to use the RRv6 versions of `useLocation` and `useParams` which meant a lot of tests had to be wrapped in Routes.

File: src/app/base/components/node/NodeTestDetails/NodeTestDetails.test.tsx [hunk 4]
```diff
@@ -47,10 +47,12 @@ describe("NodeTestDetails", () => {
...
+              <Route
+                element={<NodeTestDetails getReturnPath={getReturnPath} />}
+                path="/machine/:id/testing/:scriptResultId/details"
```
Comment (huwshimi) [inline src/app/base/components/node/NodeTestDetails/NodeTestDetails.test.tsx:53]: RRv6 `useParams` needs to have the route set up to be able to capture the params so a lot of tests had to be updated to include `<Route>`

File: src/app/devices/views/Devices.tsx [hunk 5]
```diff
@@ -1,30 +0,0 @@
-import { Route, Switch } from "react-router-dom";
-
-import DeviceDetails from "./DeviceDetails";
-import DeviceList from "./DeviceList";
-
-import NotFound from "app/base/views/NotFound";
-import devicesURLs from "app/devices/urls";
-
-const Devices = (): JSX.Element => {
```
Comment (huwshimi) [inline src/app/devices/views/Devices.tsx:9]: These routes were moved to the top level Routes file.

File: src/app/kvm/components/LXDHostVMs/LXDHostVMs.tsx [hunk 6]
```diff
@@ -54,7 +55,7 @@ const LXDHostVMs = ({
 
   if (pod) {
     return (
-      <>
+      <div {...wrapperProps}>
```
Comment (huwshimi) [inline src/app/kvm/components/LXDHostVMs/LXDHostVMs.tsx:58]: There are a number of components that had little updates to support the React Testing Library tests.

File: src/app/kvm/views/LXDClusterDetails/LXDClusterDetailsRedirect/LXDClusterDetailsRedirect.tsx [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+      <ModelNotFound
+        id={hostId}
+        inSection={false}
+        linkText="View all LXD hosts in this cluster"
+        linkURL={kvmURLs.lxd.cluster.hosts({ clusterId })}
+        modelName="LXD host"
+      />
+    );
+  }
+  return <Redirect to={kvmURLs.lxd.cluster.host.edit({ clusterId, hostId })} />;
```
Comment (huwshimi) [inline src/app/kvm/views/LXDClusterDetails/LXDClusterDetailsRedirect/LXDClusterDetailsRedirect.tsx:42]: This redirect had to be moved into its own file as it needed to be under a route for `useParams` to work.

File: src/app/kvm/views/LXDClusterDetails/LXDClusterHostSettings/LXDClusterHostSettings.tsx [hunk 8]
```diff
@@ -4,25 +4,33 @@ import { Spinner, Strip } from "@canonical/react-components";
...
 
-const LXDClusterHostSettings = ({ clusterId, hostId }: Props): JSX.Element => {
+export enum Label {
+  Loading = "Loading LXD host",
+  Title = "LXD cluster host settings",
+}
+
+const LXDClusterHostSettings = ({ clusterId }: Props): JSX.Element => {
   const navigate = useNavigate();
+  const hostId = useGetURLId(PodMeta.PK, "hostId");
```
Comment (huwshimi) [inline src/app/kvm/views/LXDClusterDetails/LXDClusterHostSettings/LXDClusterHostSettings.tsx:33]: Previously these ids were passed in from the routes file, but RRv6 needs `useParams` (which is used inside `useGetURLId`) to be called inside a route for the params to be captured.

File: src/testing/utils.tsx [hunk 9]
```diff
@@ -102,21 +102,36 @@ export const submitFormikForm = (
...
 
+  const route = <Route element={children} path={routePattern} />;
   return (
     <Provider store={getMockStore(state || rootStateFactory())}>
       <BrowserRouter>
-        <CompatRouter>{children}</CompatRouter>
+        <CompatRouter>
+          {routePattern ? (
+            <Routes>
+              {parentRoute ? <Route path={parentRoute}>{route}</Route> : route}
```
Comment (huwshimi) [inline src/testing/utils.tsx:129]: Components need to be rendered inside a Route for some hooks to work e.g. `useParams`.

File: src/app/kvm/views/LXDClusterDetails/LXDClusterHosts/LXDClusterHosts.tsx [hunk 10]
```diff
@@ -59,7 +63,7 @@ const LXDClusterHosts = ({
   );
 
   return (
-    <>
+    <div aria-label={Label.Title}>
```
Comment (petermakowski) [inline src/app/kvm/views/LXDClusterDetails/LXDClusterHosts/LXDClusterHosts.tsx:66]: is there a reason why this couldn't be a `section`? I don't think we have any styles attached to the `section` itself?

File: src/app/kvm/views/LXDSingleDetails/LXDSingleDetails.test.tsx [hunk 11]
```diff
@@ -1,21 +1,22 @@
-import { mount } from "enzyme";
-import { Provider } from "react-redux";
-import { MemoryRouter, Route } from "react-router-dom";
-import { CompatRouter } from "react-router-dom-v5-compat";
-import configureStore from "redux-mock-store";
+import { screen } from "@testing-library/react";
```
Comment (petermakowski) [inline src/app/kvm/views/LXDSingleDetails/LXDSingleDetails.test.tsx:1]: Thanks for doing this 🙏

**Review by petermakowski — APPROVED** [review]
Just a few minor comments/questions, but LGTM

**Comment by huwshimi** [comment]
> Just a few minor comments/questions, but LGTM

It would be great if when leaving comments that you don't also merge the PRs (unless you really need the changes if they're blocking) as it makes it rather hard to respond or fix the issues in the PR.

**Comment by petermakowski** [comment]
> It would be great if when leaving comments that you don't also merge the PRs (unless you really need the changes if they're blocking) as it makes it rather hard to respond or fix the issues in the PR.

Sure, will remember for the next time, but I needed your changes for my PR in this case: https://github.com/canonical-web-and-design/maas-ui/pull/4093

**Pull request merged**